### ISSUE TITLE
mutation.hh - consume clustering fragments moves out elements from btree while iterating over it

### ISSUE THREAD

**Issue by enedil** [scylladb/scylla#9207]
While iterating over range tombstones (function used by `mutataion::consume`), there is this fragment:
```cpp
            stop = consumer.consume(std::move(*rts_it));
            ++rts_it;
```
Moving out the `*rts_it` causes the internal btree node to remove itself from intrusive tree, thus invalidating the iterator for the end of the container, as well as skipping some other rows. One way to circumvent the problem is to make an explicit copy:
```
            stop = consumer.consume(range_tombstone(*rts_it));
```

I'm not sure however what is the correct solution here - perhaps the function expects that moving out the mutation fragment doesn't invalidate iterators. The mutation consumer that I used was the `mutation_rebuilder`. 

Offending fragment:
https://github.com/scylladb/scylla/blob/3633c077bed2662a2e66e6e2034856cc6df74923/mutation.hh#L191

**Comment by denesb** [comment]
The solution is to use the other move constructor: `range_tombstone(clustering_prefix, bound_kind, clustering_prefix, bound_kind, tonbstone)`.

**Comment by denesb** [comment]
`range_tombstone(range_tombstone&&)` is a huge landmine waiting for somebody to step in it.

**Comment by avikivity** [comment]
Isn't this the problem?

```c++
stop_iteration consume_clustering_fragments(const schema& s, mutation_partition& partition, Consumer& consumer) {
```

If you don't want `partition` changed, pass it as `const &`. If you are willing to have it changed, pass it as `&&` (and then nobody can see it was changed).

**Comment by denesb** [comment]
`mutation::consume()` is destructive, it even has a comment saying this. The problem is that the above can cause memory leaks.

**Comment by avikivity** [comment]
Ah. @xemul please look.

**Comment by xemul** [comment]
> Moving out the *rts_it causes the internal btree node to remove itself from intrusive tree

Why btree? RTS are stored in an `intrusive_set_external_comparator` aren't they?

**Comment by xemul** [comment]
> The solution is to use the other move constructor: range_tombstone(clustering_prefix, bound_kind, clustering_prefix, bound_kind, tonbstone)

@denesb , am I right that the partition is not expected to have its tree of tombstones (and rows) to be in its original state? In particular, it can be empty after this function returns. In this case it's better to do

```
    consumer.consume(*partition.row_tombstones().unlink_leftmost_without_rebalance());
```
isn't it? However, there's not `unlink_rightmost_without_rebalance()` for the reverse consume :(

And another question is the exception guarantee of the `consume_clustering_fragments` -- what is it? Currently if `consumer.consume` 
 hrows the offending rts gets lost, it's neither consumed nor left in the original partition.

**Comment by xemul** [comment]
> The solution is to use the other move constructor: `range_tombstone(clustering_prefix, bound_kind, clustering_prefix, bound_kind, tonbstone)`.

How about ++-ing the iterator before `std::move()`-ing the element? Like
```c++
    stop = consumer.consume(std::move(*(rst_it++)));
```

@enedil , is there a test/patch/whatever to see how the current code fails?

**Comment by denesb** [comment]
> > The solution is to use the other move constructor: `range_tombstone(clustering_prefix, bound_kind, clustering_prefix, bound_kind, tonbstone)`.
> 
> How about ++-ing the iterator before `std::move()`-ing the element? Like
> 
> ```c++
>     stop = consumer.consume(std::move(*(rst_it++)));
> ```
> 
> @enedil , is there a test/patch/whatever to see how the current code fails?

It doesn't solve the underlying problem. I had this code:
```C++
auto& rt = *rts_it++;
stop = consumer.consume(std::move(rt));
// leaks rt because the move above unlinks it from the container
```

**Comment by xemul** [comment]
Ah, right. It's not only about keeping the iterators valid.

**Comment by enedil** [comment]
@xemul 

> > Moving out the *rts_it causes the internal btree node to remove itself from intrusive tree
> 
> Why btree? RTS are stored in an `boost::intrusive::set` aren't they?

Yes, I didn't dig the code enough.

> How about ++-ing the iterator before `std::move()`-ing the element? Like
> 
> ```c++
>     stop = consumer.consume(std::move(*(rst_it++)));
> ```

This might rebalance the tree (which is the data structure underlying boost intrusive set), which invalidates all (?) the iterators after `rts_it`, so it isn't just the case of the element being currently processed. Maybe there are some better guarantees?

**Comment by enedil** [comment]
> @enedil , is there a test/patch/whatever to see how the current code fails?

```
diff --git a/mutation_rebuilder.hh b/mutation_rebuilder.hh
index c893cc26e..cb12c8f52 100644
--- a/mutation_rebuilder.hh
+++ b/mutation_rebuilder.hh
@@ -58,6 +58,14 @@ class mutation_rebuilder {
     mutation_opt consume_end_of_stream() {
         return mutation_opt(std::move(_m));
     }
+
+    void consume_new_partition(const dht::decorated_key& dk) {
+    }
+
+    stop_iteration consume_end_of_partition() {
+        return stop_iteration::yes;
+    }
+
 };
 
 // Builds the mutation corresponding to the next partition in the mutation fragment stream.
diff --git a/test/lib/mutation_source_test.cc b/test/lib/mutation_source_test.cc
index 099ea823f..1ac7952f4 100644
--- a/test/lib/mutation_source_test.cc
+++ b/test/lib/mutation_source_test.cc
@@ -1728,8 +1728,25 @@ void run_mutation_source_tests(populate_fn populate, bool with_partition_range_f
 }
 
 void run_mutation_source_tests(populate_fn_ex populate, bool with_partition_range_forwarding) {
-    run_mutation_reader_tests(populate, with_partition_range_forwarding);
+    run_mutation_reader_tests([populate] (schema_ptr s, const std::vector<mutation>& m, gc_clock::time_point t) -> mutation_source {
+        std::vector<mutation> mm;
+        for (const mutation& r : m) {
+            mutation_rebuilder b(r.decorated_key(), s);
+            auto res = mutation(r).consume(b, consume_in_reverse::yes);
+            mm.push_back(*res.result);
+        }
+        // This code won't run anyway. Let's satisfy the type checker.
+        return mutation_source([ms = populate(s, mm, t)] (schema_ptr s,
+                                              reader_permit permit,
+                                              const dht::partition_range& pr,
+                                              const query::partition_slice& slice,
+                                              const io_priority_class& pc,
+                                              tracing::trace_state_ptr tr,
+                                              streamed_mutation::forwarding fwd,
+                                              mutation_reader::forwarding mr_fwd) { return ms.make_reader(s, std::move(permit), pr, slice, pc, std::move(tr), fwd, mr_fwd); });
+    }, with_partition_range_forwarding);
 
+    run_mutation_reader_tests(populate, with_partition_range_forwarding);
     // ? -> v2 -> v1 -> *
     run_mutation_reader_tests([populate] (schema_ptr s, const std::vector<mutation>& m, gc_clock::time_point t) -> mutation_source {
         return mutation_source([ms = populate(s, m, t)] (schema_ptr s,
```

**Comment by xemul** [comment]
> This might rebalance the tree (which is the data structure underlying boost intrusive set), which invalidates all (?) the iterators after rts_it, so it isn't just the case of the element being currently processed. Maybe there are some better guarantees, rather than just "everything that follows might be invalidated".

I agree with Botond here. This should work like it does for the clustring row -- it creates a new temporary row from the one pointer by iterator by moving everything but the b-tree linkage into it. In case of range tombstone it now creates a new temporary rts by moving _everything_ into it _including_ the set linkage. What I still don't understand is why iterator pointing onto the moved-from range tombstone object allows `++` on it.

**Comment by xemul** [comment]
@enedil , please pick this patch, it should unblock you. However, it makes some cases of `schema_change_test` to crash, so I need more time before sending it
````patch
diff --git a/mutation.hh b/mutation.hh
index 408268874..2fa443432 100644
--- a/mutation.hh
+++ b/mutation.hh
@@ -188,7 +188,7 @@ stop_iteration consume_clustering_fragments(const schema& s, mutation_partition&
             emit_rt = rts_it != rts_end;
         }
         if (emit_rt) {
-            stop = consumer.consume(std::move(*rts_it));
+            stop = consumer.consume(range_tombstone(std::move(*rts_it), range_tombstone::without_link{}));
             ++rts_it;
         } else {
             stop = consumer.consume(clustering_row(std::move(*crs_it)));

````

**Comment by avikivity** [comment]
This likely has no effect on active branches, but backporting anyway (4.5 only vulnerable release) to be safe.
